### ISSUE TITLE
Validation Accuracy is 'High' but performance at inference is poor (translation with GGNN)

### ISSUE THREAD

**Issue by JustCunn** [OpenNMT/OpenNMT-py#1906]
Hi,

I've been trying to train NMT models on OpenNMT for a bit. On one of my models, it gets about `45-50` in training accuracy and about `64` in validation accuracy (I presume these are percentages).
When it comes to using `onmt_translate`/`python translate.py`, it performs very, very poorly (predicting `<UNK>` tokens all the time). If this is the case during training and validation, I don't understand how the model can achieve such scores in accuracy.

Here's my translating command
```
onmt_translate -model final-model_step_$m.pt \
               -batch_size 32 -batch_type tokens \
               -tgt tgt-test.txt -src src-test.txt \
               -n_best 1 -gpu 0 -output m$m/pred-test_beam1.txt \
               -replace_unk -dynamic_dict 2>&1 > m$m/translate.out
```
Also to note, during training, I save checkpoints every 50,000 steps (so maybe performance drops considerably since last val accuracy check). My target vocab is also large, so I doubt that words are getting replaced with `<UNK>` at a high rate.

I'm unsure as to whether I'm misunderstanding something, and the model had poor performance all the time, or this is a problem I need to address somewhere.

All advice is welcome, and thanks in advance.

**Comment by francoishernandez** [comment]
These accuracy figures seem quite low for a translation task. What kind of dataset are you using?
Predicting <unk> may stem from several issues. Do you apply proper tokenization to your data?

Sharing your training command/config might help as well.

**Comment by JustCunn** [comment]
@francoishernandez (apologies for the long comment) Oh, I didn't realise that was really low (I don't generally use OpenNMT). I'm using a dataset with around 650k sentences in the English-Irish pair (Irish is a very low-resourced language). 

Just to highlight, I'm using a GGNN, and so my source vocab is limited to 1000 words which I know is a constraint. I'm working on allowing that to be expanded and have already been in contact with the author of the GGNN in OpenNMT. So, that means my source files are preprocessed in a completely different way, but looking at my target files I somehow forgot to do BPE, but done normal tokenisation! I'm also working on being able to use my full corpus (~900k sentences).

Train config (some flags are GGNN specific):
```
onmt_train \
-data $data_path/final \
-encoder_type ggnn -layers 2 -decoder_type rnn -rnn_size 1024 \
-learning_rate 0.1 \
-start_decay_steps 15000 -learning_rate_decay 0.8 \
-global_attention general \
-batch_size 32 \
-word_vec_size 1024 \
-position_encoding \
-bridge -train_steps 300000 \
-accum_count 2 \
-world_size 2 -gpu_ranks 0 1 \
-save_checkpoint_steps 50000 \
-save_model $data_path/final-model \
-src_vocab $data_path/srcvocab.txt \
-n_edge_types 7 \
-state_dim 1024 \
-n_steps 10 \
-n_node 120 > $data_path/train.final.out
```
Another oversight by me is that I don't pass the target vocab list I have during training (it's used during the prprocessing step).

As a result of using a GGNN without modifications I plan to make, I always expected poorer than normal performance. Whether this helps or not, but perplexity started at around 450 and lowered around 19. Would that train/val accuracy be expected to give back basically terrible predictions?

Thanks as well for offering your help.

**Comment by francoishernandez** [comment]
I honestly have no clue of what's expected for GGNN in the translation context.
Did you first try a 'traditional' translation setup like a Transformer or RNN on the same datasets?
Also, 650k is quite low resource indeed. Maybe you could try pretraining with some similar more high resource language(s).

**Comment by JustCunn** [comment]
@francoishernandez Yes, I have multiple transformer models with about 600k of my corpus and 1000 source vocab too and upon checking them they produce ok translations, though not the best but I can make out what they're trying to say. It seems the issue lies with the current constraints of the GGNN encoder which I'll try to resolve. Thanks for guiding me through that.

Just to ask since I mentioned preprocessing (apologies if this is going off-topic), I've been using `onmt_preprocess` for all my preprocessing so far, both with the transformer and the GGNN, but it appears that's become deprecated. I pass `onmt_preprocess` my full 900k corpus, but the data file it creates only has 600k examples. Does `build_vocab` replace this? I'm only asking here because it no longer has full documentation and you seem to be quite familiar with OpenNMT.

**Comment by francoishernandez** [comment]
Yes, preprocessing was removed in 2.0: https://github.com/OpenNMT/OpenNMT-py#announcement---opennmt-py-20

You can have a look at the [updated docs](https://opennmt.net/OpenNMT-py/quickstart.html) if you want to switch to 2.0.

**Comment by JustCunn** [comment]
Thanks a lot for helping me with that. Just to say, for anyone who may have the original 'issue' too, I'm working on expanding the amount of source vocab that can be used in the GGNN.
